File: main/word-search-ii/word-search-ii.py
#!/usr/bin/env python3

BIAS = ord('a')
ALPHA_LEN = ord('z') - BIAS + 1
ALPHA = 'abcdefghijklmnopqrstuvwxyz'
assert all(ord(ALPHA[i]) - BIAS == i for i in range(ALPHA_LEN)), \
        'English lower-case letters are noncontiguous'

class CharMap(object):
    def __init__(self):
        self._elems = [None] * ALPHA_LEN

    def __getitem__(self, ch):
        return self._elems[ord(ch) - BIAS]

    def __setitem__(self, ch, val):
        self._elems[ord(ch) - BIAS] = val

class Trie(CharMap):
    def __init__(self, words=()):
        super().__init__()
        self._word = None

        for word in words:
            cur = self
            for ch in word:
                if cur[ch] is None:
                    cur[ch] = Trie()
                cur = cur[ch]
            cur._word = word

    @property
    def word(self):
        return self._word

class Solution:
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        istop = len(board)
        jstop = len(board[0])
        assert all(len(row) == jstop for row in board)
        matches = []

        def dfs(trie, i, j):
            if i == -1 or i == istop or j == -1 or j == jstop:
                return

            ch = board[i][j]
            if ch is None:
                return

            trie = trie[ch]
            if trie is None:
                return

            if trie.word is not None:
                matches.append(trie.word)

            board[i][j] = None
            dfs(trie, i, j - 1)
            dfs(trie, i, j + 1)
            dfs(trie, i - 1, j)
            dfs(trie, i + 1, j)
            board[i][j] = ch

        root = Trie(words)
        for i in range(istop):
            for j in range(jstop):
                dfs(root, i, j)

        return list(set(matches))

File: main/word-search-ii/word-search-ii.cs
internal class CharMap<T> { // maps from a lower-case English letter to a T
    public const string Alpha = "abcdefghijklmnopqrstuvwxyz";

    private const int AlphaLen = 'z' - 'a' + 1;

    static CharMap() {
        for (var i = 0; i != AlphaLen; ++i) {
            if (Alpha[i] - 'a' != i) {
                throw new NotSupportedException(
                        "English lower-case letters are noncontiguous");
            }
        }
    }

    public T this[char ch] {
        get => _elems[ch - 'a'];
        set => _elems[ch - 'a'] = value;
    }

    private T[] _elems = new T[AlphaLen];
}

internal sealed class Trie : CharMap<Trie> {
    public string Word { get; private set; }

    public Trie(params string[] words) {
        foreach (var word in words) {
            var cur = this;
            foreach (var ch in word) cur = cur[ch] ?? (cur[ch] = new Trie());
            cur.Word = word;
        }
    }

#if WITH_DUMP_LOGIC
    private object ToDump() {
        IDictionary<string, object> repr = new System.Dynamic.ExpandoObject();

        foreach (var ch in Alpha) {
            var child = this[ch];
            if (child != null) repr[ch.ToString()] = child;
        }

        if (Word != null) repr[nameof(Word)] = Word;

        return repr;
    }
#endif
}

public class Solution {
    public IList<string> FindWords(char[,] board, string[] words) {
        const char visited = '\0';
        var istop = board.GetLength(0);
        var jstop = board.GetLength(1);
        var matches = new List<string>();

        void Dfs(Trie trie, int i, int j) {
            if (i == -1 || i == istop || j == -1 || j == jstop) return;

            var ch = board[i, j];
            if (ch == visited) return;

            trie = trie[ch];
            if (trie == null) return;

            if (trie.Word != null) matches.Add(trie.Word);

            board[i, j] = visited;
            Dfs(trie, i, j - 1);
            Dfs(trie, i, j + 1);
            Dfs(trie, i - 1, j);
            Dfs(trie, i + 1, j);
            board[i, j] = ch;
        }

        var root = new Trie(words);
        for (var i = 0; i != istop; ++i)
            for (var j = 0; j != jstop; ++j) Dfs(root, i, j);

        return matches.Distinct().ToList();
    }
}
